### PR TITLE
2220 Decollate batch into list after iteration computation (29/June)

### LINKED ISSUE #2220: Transforms need to handle list of channel-first data
**Is your feature request related to a problem? Please describe.**
Currently, if using the default collate_fn, the `interted` data is a list of channel-first Tensor or numpy array, if we want to apply transforms on the data, most of the MONAI transforms can't support it.
For example, users want to use `CopyItemsd`, `AsDiscreted`, `KeepLargestConnectedComponentd`, etc.
All these transforms can only support **a batch-first Tensor** or **a channel-first numpy array**.

Thanks.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2220 #2300 #2275 #2410 #2452 .

### Description
This PR added support to decollate the iteration output.
The main idea is to decollate the output dict / list of an iteration into a list of dicts / lists, copy the scalar values(like loss) into every dict / list. then all the following logic are based on this list of dicts / lists.

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @ericspod @wyli @rijobro ,

This PR is ready for review now, implementation is based on our previous online discussion.
Please note that there is a minor drawback of this idea: the `batch_transform` and `output_transform` callables can be more complicated if users don't use `from_engine()` util to extract the `pred`, `label` or `image`.

Thanks.

**Comment by Nic-Ma** [comment]
When ignite updated this Friday, it will support a list of tensors in metrics:
https://github.com/pytorch/ignite/pull/2055#issuecomment-862360315

Thanks.

File: monai/data/csv_saver.py [hunk 1]
```diff
@@ -87,11 +87,12 @@ def save(self, data: Union[torch.Tensor, np.ndarray], meta_data: Optional[Dict]
...
+        data_: np.ndarray
         if isinstance(data, torch.Tensor):
-            data = data.detach().cpu().numpy()
-        if not isinstance(data, np.ndarray):
-            raise AssertionError
-        self._cache_dict[save_key] = data.astype(np.float32)
+            data_ = data.detach().cpu().numpy()
+        else:
+            data_ = np.asarray(data)
+        self._cache_dict[save_key] = data_.astype(np.float32)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/data/csv_saver.py:95]: updated in the latest commit.
Thanks.

File: monai/engines/utils.py [hunk 2]
```diff
@@ -133,7 +133,7 @@ def default_make_latent(
 
 def engine_apply_transform(batch: Any, output: Any, transform: Callable):
     """
-    Apply transform for the engine.state.batch and engine.state.output.
+    Apply transform for one item of the engine.state.batch and engine.state.output.
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/engines/utils.py:136]: Sure, updated in the latest commit.
Thanks.

File: monai/engines/workflow.py [hunk 3]
```diff
@@ -146,26 +154,82 @@ def set_sampler_epoch(engine: Engine):
             else:
                 raise ValueError("event_names must be a list or string or EventEnum.")
 
+        self._register_decollate()
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/engines/workflow.py:157]: Sure, added to all the workflows in the latest commit.
Thanks.

File: monai/handlers/classification_saver.py [hunk 4]
```diff
@@ -108,14 +102,18 @@ def __call__(self, engine: Engine) -> None:
...
-        filenames = self.batch_transform(engine.state.batch).get(Key.FILENAME_OR_OBJ)
-        if issequenceiterable(filenames):
-            self._filenames.extend(filenames)
-        outputs = self.output_transform(engine.state.output)
-        if outputs is not None:
-            if isinstance(outputs, torch.Tensor):
-                outputs = outputs.detach()
-            self._outputs.append(outputs)
+        meta_data = self.batch_transform(engine.state.batch)
+        if isinstance(meta_data, dict):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/handlers/classification_saver.py:106]: Hi @wyli ,

`meta_data` can be a `dict` if decollate=False or `list of dict` if decollate=True.
So try to decollate it if it's a `dict`.

Thanks.

File: monai/handlers/metrics_saver.py [hunk 5]
```diff
@@ -105,9 +106,14 @@ def _started(self, engine: Engine) -> None:
...
-            if issequenceiterable(filenames):
-                self._filenames.extend(filenames)
+            meta_data = self.batch_transform(engine.state.batch)
+            if isinstance(meta_data, dict):
+                # decollate the `dictionary of list` to `list of dictionaries`
+                meta_data = decollate_batch(meta_data)
+            for m in meta_data:
+                filename = m.get(Key.FILENAME_OR_OBJ)
+                if filename is not None:
+                    self._filenames.append(filename)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/handlers/metrics_saver.py:116]: Sure, updated in the latest commit.
Thanks.

File: monai/handlers/classification_saver.py [hunk 6]
```diff
@@ -108,14 +102,18 @@ def __call__(self, engine: Engine) -> None:
...
+            # decollate the `dictionary of list` to `list of dictionaries`
+            meta_data = decollate_batch(meta_data)
+        engine_output = self.output_transform(engine.state.output)
+        for m, o in zip(meta_data, engine_output):
+            filename = m.get(Key.FILENAME_OR_OBJ)
+            if filename is not None:
+                self._filenames.append(filename)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/handlers/classification_saver.py:113]: updated in the latest commit.
Thanks.

File: monai/transforms/post/dictionary.py [hunk 7]
```diff
@@ -340,26 +334,6 @@ def __init__(
         super().__init__(keys, ensemble, output_key)
 
 
-class Decollated(MapTransform):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/post/dictionary.py:343]: Hi @wyli ,

I think now we suppose all the regular transforms should work on `channel-first` data, so no need to make a transform chain with `Decollated` transform.
if users want to decollate a batch data, they can use the `monai.transforms.BatchInverseTransform`.
What do you think?

Thanks.

File: monai/transforms/utility/array.py [hunk 8]
```diff
@@ -241,36 +241,24 @@ class SplitChannel(Transform):
...
     Split Numpy array or PyTorch Tensor data according to the channel dim.
     It can help applying different following transforms to different channels.
-    Channel number must be greater than 1.
 
     Args:
-        channel_dim: which dimension of input image is the channel, default to None
-            to automatically select: if data is numpy array, channel_dim is 0 as
-            `numpy array` is used in the pre transforms, if PyTorch Tensor, channel_dim
-            is 1 as in most of the cases `Tensor` is uses in the post transforms.
+        channel_dim: which dimension of input image is the channel, default to 0.
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/utility/array.py:246]: Hi @wyli @rijobro , this PR already solved the issue in #2275 , the test program in that issue is not correct, should use `deepcopy` to create another Tensor data for test. Here is some test result:
```
>>> import numpy as np
>>> import torch
>>> from monai.transforms import SplitChannel
>>> 
>>> from copy import deepcopy
>>> 
>>> t = SplitChannel()
>>> 
>>> _np = np.random.randint(2, size=(3, 3, 4))
>>> _torch = torch.Tensor(deepcopy(_np))
>>> 
>>> print(t(_np)[0].shape)
(1, 3, 4)
>>> print(t(_torch)[0].shape)
torch.Size([1, 3, 4])
```

Thanks.

File: monai/utils/jupyter_utils.py [hunk 9]
```diff
@@ -143,13 +143,13 @@ def tensor_to_images(name: str, tensor: torch.Tensor):
     color channels RGB or RGBA. This allows multiple images to be created from a single tensor, ie. to show
     each channel separately.
     """
+    if tensor.ndim == 3 and tensor.shape[1] > 2 and tensor.shape[2] > 2:
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/utils/jupyter_utils.py:146]: Hi @wyli ,

Here we changed the previous `batch-first` data to `channel-first` single image, so I changed from `tensor.ndim = 4 or 5` to `3 or 4`. May I know why you think to support `tensor.ndim == 2`?

Thanks.

File: monai/engines/workflow.py [hunk 10]
```diff
@@ -146,26 +154,82 @@ def set_sampler_epoch(engine: Engine):
...
+                        return v.shape[0]
+                for v in batch_data:
+                    if issequenceiterable(v):
+                        warnings.warn(
+                            "batch_data doesn't contain batched Tensor data, use the length of first sequence data."
+                        )
+                        return len(v)
+                raise RuntimeError("failed to automatically detect the batch size.")
+
+            def _copy_scalar(batch_data):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/engines/workflow.py]: Sure, I converted to a utility in the latest commit.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli @ericspod ,

Thanks for your detailed reviews.
I updated the PR according to your comments.
And changed all the pre transforms and post transforms to `pre_processing` and `post_processing`.
Could you please help review it again?

Thanks.

**Comment by Nic-Ma** [comment]
Mark: when PR https://github.com/Project-MONAI/MONAI/pull/2451 merged, I will update the integration test that computes a list tensors in ignite metrics.

Thanks.

File: monai/data/dataset.py [hunk 11]
```diff
@@ -173,7 +173,7 @@ def __init__(
             if not self.cache_dir.is_dir():
                 raise ValueError("cache_dir must be a directory.")
 
-    def _pre_transform(self, item_transformed):
+    def _preprocessing(self, item_transformed):
```
Comment (wyli) [inline monai/data/dataset.py:176]: I think the previous usage of `pre_transform` here is correct. https://github.com/Project-MONAI/MONAI/issues/2410 is mainly about the workflows.

File: monai/engines/trainer.py [hunk 12]
```diff
@@ -85,6 +85,8 @@ class SupervisedTrainer(Trainer):
             new events can be a list of str or `ignite.engine.events.EventEnum`.
         event_to_attr: a dictionary to map an event to a state attribute, then add to `engine.state`.
             for more details, check: https://github.com/pytorch/ignite/blob/v0.4.4.post1/ignite/engine/engine.py#L160
+        decollate: whether decollate the batch-first data to a list of data after model computation, default to `True`.
```
Comment (wyli) [inline monai/engines/trainer.py:88]:
```suggestion
        decollate: whether to decollate the batch-first data to a list of data after model computation, default to `True`.
```

File: monai/engines/trainer.py [hunk 13]
```diff
@@ -85,6 +85,8 @@ class SupervisedTrainer(Trainer):
...
+            if `False`, postprocessing may not work as all the transforms should apply on channel-first data.
```
Comment (wyli) [inline monai/engines/trainer.py:89]:
```suggestion
            if `False`, postprocessing may not work as the `monai.transforms` module assumes channel-first data.
```

File: monai/engines/workflow.py [hunk 14]
```diff
@@ -146,26 +155,40 @@ def set_sampler_epoch(engine: Engine):
...
-    def _register_post_transforms(self, posttrans: Callable):
+    def _register_decollate(self):
+        """
+        Register the decollate operation for batch data, will execure after model forward and loss forward.
+
+        """
+
+        @self.on(IterationEvents.MODEL_COMPLETED)
+        def _decollate_data(engine: Engine) -> None:
+            if isinstance(engine.state.batch, (dict, list)):
```
Comment (wyli) [inline monai/engines/workflow.py:175]: should this `isinstance` be a part of `copy_scalar_to_batch`? I think `copy_scalar_to_batch` could be `replicate_scalar_to_batch` or `rep_scalar_to_batch`

File: monai/handlers/post_processing.py [hunk 15]
```diff
@@ -23,16 +23,16 @@
 
 class PostProcessing:
     """
```
Comment (wyli) [inline monai/handlers/post_processing.py:25]: the file name could be `postprocessing.py`

File: monai/handlers/post_processing.py [hunk 16]
```diff
@@ -49,8 +49,5 @@ def __call__(self, engine: Engine) -> None:
         Args:
             engine: Ignite Engine, it can be a trainer, validator or evaluator.
         """
-        engine.state.batch, engine.state.output = engine_apply_transform(
-            batch=engine.state.batch,
-            output=engine.state.output,
-            transform=self.transform,
-        )
+        for i, (b, o) in enumerate(zip(engine.state.batch, engine.state.output)):
```
Comment (wyli) [inline monai/handlers/post_processing.py:52]: what's the type of `engine.state.batch` and `output`? probably need a check here

File: monai/handlers/stats_handler.py [hunk 17]
```diff
@@ -174,7 +174,7 @@ def _default_iteration_print(self, engine: Engine) -> None:
         """
         Execute iteration log operation based on Ignite engine.state data.
         Print the values from Ignite state.logs dict.
-        Default behavior is to print loss from output[1], skip if output[1] is not loss.
+        Default behavior is to print loss from output[0] as output is decollated list.
```
Comment (wyli) [inline monai/handlers/stats_handler.py:177]: I think we need some additional documentation to explain this...
```suggestion
        The default behavior is to print loss from output[0] as output is a decollated list.
```

**Review by wyli — COMMENTED** [review]
thanks, please see some comments inline

File: tests/test_label_to_contour.py [hunk 18]
```diff
@@ -147,30 +147,33 @@ def test_contour(self):
...
+            test_result_np = test_result_img.cpu().numpy()
             for channel in range(channels):
-                np.testing.assert_allclose(test_result_img[batch, channel, ...], expected_output)
+                np.testing.assert_allclose(test_result_np[channel, ...], expected_output)
 
         # check invalid input data
-        error_input = torch.rand(1, 2, 3)
+        error_input = torch.rand(
+            1,
+            2,
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply tests/test_label_to_contour.py:173]: I feel this is some new `codeformat` feature in `/black`.

File: monai/utils/jupyter_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wyli) [reply monai/utils/jupyter_utils.py:146]: ok looks like line 194 `for src in (engine.state.batch, engine.state.output)` is incorrect

File: monai/transforms/post/dictionary.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/post/dictionary.py:343]: Sure, added back to `inverse_batch_transform.py` as it has different input data shape with other post transforms.

Thanks.

File: tests/test_label_to_contour.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply tests/test_label_to_contour.py:173]: updated in the latest commit.
Thanks.

File: monai/data/dataset.py (hunk 11 above)
Comment (Nic-Ma) [reply monai/data/dataset.py:176]: Sure, restored all the `_pre_transform` and `_post_transform` in datasets in the latest commit.

Thanks.

File: monai/engines/trainer.py (hunk 12 above)
Comment (Nic-Ma) [reply monai/engines/trainer.py:88]: updated all the same doc-strings in workflows.

Thanks.

File: monai/engines/trainer.py (hunk 13 above)
Comment (Nic-Ma) [reply monai/engines/trainer.py:89]: updated all the doc-strings in workflows.

Thanks.

File: monai/engines/workflow.py (hunk 14 above)
Comment (Nic-Ma) [reply monai/engines/workflow.py:175]: Changed to `rep_scalar_to_batch` to align with `ensure_tuple_rep` in the latest commit.

Thanks.

File: monai/handlers/post_processing.py (hunk 15 above)
Comment (Nic-Ma) [reply monai/handlers/post_processing.py:25]: updated in the latest commit.
Thanks.

File: monai/handlers/stats_handler.py (hunk 17 above)
Comment (Nic-Ma) [reply monai/handlers/stats_handler.py:177]: enhanced the doc-strings in the latest commit.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I updated the integration test to use ignite v0.4.5 metrics, hopefully, it will work when the ignite 0.4.5 PR merged.

Thanks.

**Comment by wyli** [comment]
> Hi @wyli ,
> 
> I updated the integration test to use ignite v0.4.5 metrics, hopefully, it will work when the ignite 0.4.5 PR merged.
> So I canceled the CI tests now, let's test later.
> 
> Thanks.

sure, I'll run all the nightly tests on this branch before merging

**Comment by Nic-Ma** [comment]
Thanks, let me fix all the merge conflicts.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your quick fix in CI config, added the issue to this PR description.

**Review by wyli — APPROVED** [review]
thanks! I did my best to review this PR, and all the CI pipelines seem to be happy with it.
